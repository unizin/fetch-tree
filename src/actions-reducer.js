export const ACTION_DONE = '@@fetchTree/action_done'
export const STORE_ENTITY = '@@fetchTree/store_entity'

export const setActionStatusDone = (id) => ({
    type: ACTION_DONE,
    payload: id,
})

export const storeEntity = (id, value) => ({
    type: STORE_ENTITY,
    payload: { id, value },
})

export const selectEntity = (state, id) => (
    state.fetchTree != null
    && state.fetchTree.entities != null
    && state.fetchTree.entities[id]
)

export const selectIsReady = (state, id) => Boolean(
    state.fetchTree != null
    && state.fetchTree.ready != null
    && state.fetchTree.ready[id] != null
)


function doneReducer(state = {}, action) {
    if (action.type === 'ACTION_DONE') {
        return {
            ...state,
            [action.payload]: true,
        }
    }
    return state
}

function entityReducer(state = {}, action) {
    if (action.type === STORE_ENTITY) {
        const { id, value } = action.payload
        return {
            ...state,
            [id]: value,
        }
    }
    return state
}


const defaultState = {
    ready: doneReducer(undefined, { type: '__init__' }),
    entities: entityReducer(undefined, { type: '__init__' }),
}

export default function reducer(state = defaultState, action) {
    switch (action.type) {
        case ACTION_DONE:
            return {
                ...state,
                done: doneReducer(state.done, action),
            }
        case STORE_ENTITY:
            return {
                ...state,
                entities: entityReducer(state.done, action),
            }

        default:
            return state
    }
}
