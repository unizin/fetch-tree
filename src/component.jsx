import React from 'react'
import { connect } from 'react-redux'
import loaderContext from './loader_context'
import processor from './processor'
import withProps from './node_types/with-props'
import withDispatch from './node_types/with-dispatch'

const getDisplayName = Component => (
  Component.displayName ||
  Component.name ||
  (typeof Component === 'string' ? Component : 'Component')
)


const contextShape = React.PropTypes.shape({
    execute: React.PropTypes.func.isRequired,
    hasExecuted: React.PropTypes.func.isRequired,
})

// These need to get passed as props and we need to avoid name conflicts with
// any props the user is passing through to Component. Using the dashes and
// colons ensures that we can't accidentally conflict with JSX props.
const IS_READY = '--loader:is-ready'
const ACTION_QUEUE = '--loader:action-queue'
const DISPATCH = '--loader:dispatch'
const DISPATCH_PROXY = '--loader:dispatchProxy'

export default function ({ component: Component, busy: Busy, resources, mapDispatchToProps }) {
    if (mapDispatchToProps) {
        throw new Error('mapDispatchToProps was removed. Use the dispatch node type instead')
    }
    const displayName = `Loader(${getDisplayName(Component)})`

    if (resources.TYPE !== 'group') {
        if (resources.TYPE === 'debug' && resources.child.TYPE === 'group') {
            // It's ok to wrap the group in a debug node
        } else {
            throw new Error('resources must be a `group()`')
        }
    }

    class LoaderComponent extends React.Component {
        static displayName = displayName

        static childContextTypes = {
            loaderContext: contextShape,
        }

        static contextTypes = {
            loaderContext: contextShape,
        }

        constructor(props, context) {
            super(props, context)
            this.loaderContext = loaderContext(context.loaderContext)
            const {
                [DISPATCH]: dispatch,
                [DISPATCH_PROXY]: dispatchProxy,
            } = props
            dispatchProxy.dispatch = dispatch
        }

        getChildContext() {
            return {
                loaderContext: this.loaderContext,
            }
        }

        componentDidMount() {
            const {
                [ACTION_QUEUE]: actionQueue,
                [DISPATCH]: dispatch,
            } = this.props
            this.loaderContext.execute(dispatch, actionQueue)
        }

        componentWillReceiveProps(nextProps) {
            const {
                [DISPATCH]: dispatch,
                [DISPATCH_PROXY]: dispatchProxy,
            } = nextProps
            dispatchProxy.dispatch = dispatch
        }

        componentDidUpdate() {
            const {
                [ACTION_QUEUE]: actionQueue,
                [DISPATCH]: dispatch,
            } = this.props
            this.loaderContext.execute(dispatch, actionQueue)
        }

        render() {
            const {
                [IS_READY]: isReady,
                [ACTION_QUEUE]: ignoredActionQueue,
                [DISPATCH_PROXY]: ignoredDispatchProxy,
                ...props,
            } = this.props

            if (!isReady) {
                if (Busy) {
                    return (<Busy />)
                }

                return (<noscript></noscript>)
            }

            return (
                <Component {...props} />
            )
        }
    }

    function mapStateToProps() {
        // Each instance gets its own `dispatchProxy` function that never changes
        const dispatchProxy = (action) => {
            // dispatchProxy.dispatch will get attached in the component's constructor
            // and updated in componentWillReceiveProps
            if (!dispatchProxy.dispatch) {
                throw new Error(`Dispatch can't be called yet`)
            }
            return dispatchProxy.dispatch(action)
        }

        return (state, props) => {
            const localResources = withProps(props,
                withDispatch(dispatchProxy, resources)
            )
            const { isReady, actionQueue, value } = processor(localResources, state)

            return {
                ...value,
                [IS_READY]: isReady,
                [ACTION_QUEUE]: actionQueue,
                [DISPATCH_PROXY]: dispatchProxy,
            }
        }
    }

    const mapDispatch = {
        [DISPATCH]: (action) => action,
    }

    return connect(mapStateToProps, mapDispatch)(LoaderComponent)
}
