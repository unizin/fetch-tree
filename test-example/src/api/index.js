import * as api from './api'

let counts = {}

Object.keys(api).reduce((api, k) => {
    const implementation = api[k]

    api[k] = function apiCounter(...args) {
        counts[k] = 1 + (counts[k] || 0)
        return implementation.apply(this, args)
    }

    return api
}, api)


export const resetCounts = () => { counts = {} }
export const getCounts = () => counts

export default api
