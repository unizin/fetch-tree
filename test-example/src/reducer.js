import { combineReducers } from 'redux'
import { reducer as fetchTree } from 'fetch-tree'

function todos(state = {}, action) {
    switch (action.type) {
        case 'SET_TODO': {
            return {
                ...state,
                [action.meta.id]: action.payload,
            }
        }
        default:
            return state
    }
}

function count(state = 0, action) {
    switch (action.type) {
        case 'INCREMENT_COUNT':
            return state + 1
        default:
            return state
    }
}

export default combineReducers({
    currentUsername: 'bob',
    todos,
    count,
    fetchTree,
})


export const selectTodo = (state, id) => state.todos[id]

export const selectCount = (state) => state.count

export const selectCurrentUsername = (state) => state.currentUsername
