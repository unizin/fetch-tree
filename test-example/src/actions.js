import api from './api'

export const fetchTodo = (id) => (dispatch) => {
    return api.fetchTodo(id).then(todo => {
        dispatch({
            type: 'SET_TODO',
            payload: todo,
            meta: { id },
        })
    })
}

export const incrementCount = () => ({
    type: 'INCREMENT_COUNT',
})
