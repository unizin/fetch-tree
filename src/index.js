import group from './node_types/group'
import loader from './node_types/loader'
import selector from './node_types/selector'
import depends from './node_types/depends'
import debug from './node_types/debug'
import virtual from './node_types/virtual'
import entity from './node_types/entity'
import withContext from './node_types/with-context'
import child from './node_types/child'
import withProps from './node_types/with-props'
import withDispatch from './node_types/with-dispatch'
import dispatch from './node_types/dispatch'
import fromProps from './node_types/from-props'

import component from './component'
import reducer, { refreshCache } from './actions-reducer'

import { register as registerNodeType } from './processor.js'

export { group, loader, selector, depends, reducer, debug, virtual, entity, withContext }
export { withDispatch, dispatch }
export { child, withProps, fromProps }
export { refreshCache }

export { registerNodeType }
export default component
