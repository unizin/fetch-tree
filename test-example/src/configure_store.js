import { createStore, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'
import reducer from './reducer'

export default function configureStore() {
    const store = createStore(
        reducer,
        undefined,
        applyMiddleware(
            thunk
        )
    )

    return store
}
