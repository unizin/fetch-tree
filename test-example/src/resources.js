import { loader, entity } from 'fetch-tree'
import { fetchTodo } from './actions'
import { selectTodo } from './reducer'
import api from './api'

export const todoResource = loader({
    id: (id) => `todo-${id}`,
    action: fetchTodo,
    selector: selectTodo,
})

export const branchResource = entity({
    id: (id) => `branch-${id}`,
    apiFunction: api.fetchBranch,
})

export const userResource = entity({
    id: (id) => `user-${id}`,
    apiFunction: api.fetchUser,
})
