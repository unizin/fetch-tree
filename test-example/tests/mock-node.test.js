import React from 'react'
import td from 'testdouble'

import { renderSnapshot } from './test-utils.js'
import fetchTree, { group, fromProps, virtual, child, mockNode, resetMocks, dispatch, registerNodeType } from 'fetch-tree'

import Leaf from '../src/components/todo'
import Loading from '../src/components/loading'
import { todoResource } from '../src/resources'

afterEach(resetMocks)

// This doesn't need a resourceGroup because it will use the one defined in Leaf
const ConnectedLeaf = fetchTree({
    component: Leaf,
    busy: Loading,
})

function Root(props) {
    return (
        <ul>
            {props.ids.map(id => (
                <li key={id}>
                    <ConnectedLeaf id={id} />
                </li>
            ))}
        </ul>
    )
}

const ConnectedRoot = fetchTree({
    component: Root,
    resourceGroup: group({
        // virtual allows you to define nodes that don't get added to the props
        idsForThisComponent: virtual(fromProps('ids')),
        preloadLeaves: virtual(child(
            ['idsForThisComponent'],
            // While `ConnectedRoot` is holding a loading screen up, it loads
            // the data for the connected children.
            (ids) => ids.map(id => (
                <ConnectedLeaf id={id} />
            ))
        )),

    }),
})

test(`mockNode() will replace the way a node processes`, async () => {
    mockNode(todoResource, (id) => {
        let note
        // You can check your parameters any way you want.
        if (id === 1) {
            note = 'one'
        } else if (id === 2) {
            note = 'two'
        } else {
            throw new Error('')
        }
        return {
            id,
            done: true,
            note,
        }
    })
    const report = await renderSnapshot(
        <ConnectedRoot ids={[1, 2]} />
    )

    expect(report.apiRequests).toEqual({})
    // Nothing needed to be fetched, so it was able to render immediately.
    expect(report.loadingScreens).toBe(0)
})


test(`resetMocks())`, async () => {
    mockNode(todoResource, (id) => {
        return { id, done: true, note: 'whatever' }
    })

    resetMocks()
    const report = await renderSnapshot(
        <ConnectedRoot ids={[1, 2]} />
    )

    expect(report.apiRequests).toEqual({
        fetchTodo: 2,
    })
    expect(report.loadingScreens).toBe(1)
})

test(`mocking a mocked node replaces it`, async () => {
    const mock1 = td.function('.mock1')
    const mock2 = td.function('.mock2')
    td.when(mock2(1)).thenReturn({
        id: 1, done: true, note: 'second mock',
    })

    mockNode(todoResource, mock1)
    mockNode(todoResource, mock2)

    const report = await renderSnapshot(
        <ConnectedRoot ids={[1]} />
    )

    td.verify(mock1(), { ignoreExtraArgs: true, times: 0 })

    expect(report.apiRequests).toEqual({})
    // Nothing needed to be fetched, so it was able to render immediately.
    expect(report.loadingScreens).toBe(0)
})

describe(`If you mock anything, you must mock:`, () => {
    async function captureError(component) {
        try {
            await renderSnapshot(component)
        } catch (e) {
            return e
        }
        return undefined
    }
    const Dummy = () => (<div>Dummy</div>)
    beforeEach(() => {
        // unused mock
        mockNode(todoResource, (id) => {
            return { id, done: true, note: 'whatever' }
        })
    })

    async function testError(resourceGroup, regex) {
        const Component = fetchTree({
            component: Dummy,
            resourceGroup,
        })

        const error = await captureError(
            <Component ids={[1]} />
        )
        expect(error).not.toBeUndefined()
        expect(error.message).toMatch(regex)
    }


    test(`Any async node`, async () => {
        const TYPE = 'never-ready'
        const neverReady = registerNodeType({
            TYPE,
            factory() { return { TYPE } },
            nodeProcessor() {
                return {
                    isReady: false,
                }
            },
        })

        await testError(group({
            whatever: neverReady(),
        }), /^MissingMock:/)
    })

    test(`dispatch nodes`, async () => {
        const actionCreator = () => ({
            type: 'something',
        })
        await testError(group({
            doSomething: dispatch(actionCreator),
        }), /^MissingMock: dispatch/)
    })

    test(`select nodes`, async () => {
        function selectFoo(state) { return state.foo }

        await testError(group({
            foo: selectFoo,
        }), /^MissingMock: dispatch/)
    })
})
