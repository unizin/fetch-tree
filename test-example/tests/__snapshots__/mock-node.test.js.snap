exports[`test mockNode() will replace the way a node processes 1`] = `
<ul>
  <li>
    <div>
      id: 1
      <input
        checked={true}
        type="checkbox" />
      <p>
        one
      </p>
    </div>
  </li>
  <li>
    <div>
      id: 2
      <input
        checked={true}
        type="checkbox" />
      <p>
        two
      </p>
    </div>
  </li>
</ul>
`;

exports[`test mocking a mocked node replaces it 1`] = `
<ul>
  <li>
    <div>
      id: 1
      <input
        checked={true}
        type="checkbox" />
      <p>
        second mock
      </p>
    </div>
  </li>
</ul>
`;

exports[`test resetMocks()) 1`] = `
<ul>
  <li>
    <div>
      id: 1
      <input
        checked={false}
        type="checkbox" />
      <p>
        Fouti hi rovumjak wuwab ojki utemun ti voblom mefup ivkiz nadahuw ci.
      </p>
    </div>
  </li>
  <li>
    <div>
      id: 2
      <input
        checked={true}
        type="checkbox" />
      <p>
        Bev rip asu rosob lo su za wirun kubvaw nadridzez cabfed temekuinu co.
      </p>
    </div>
  </li>
</ul>
`;
